feat(gux-card): addition of new card component

gavin-everett-genesys opened the pull request:
Card Container Component.

katie-bobbe-genesys commented on src/components/beta/gux-card/gux-card.tsx:
```diff
@@ -0,0 +1,47 @@
+          <h1 class="gux-card-header">
+            <slot name="title" />
+          </h1>
+          <div class="gux-card-content">
```
Does the card component need `title` and `footer` slots? Or should the card just function as a container for the default slot?

gavin-everett-genesys replied:
Yep you are correct, it doesn't state to have title and footer slots so ill adjust to just have a default slot. I added them in  for structure but from the design just looks like a flexible container. Currently awaiting an updated design file so I will check. @katie-bobbe-genesys

katie-bobbe-genesys commented on src/components/beta/gux-card/tests/gux-card.e2e.ts:
```diff
@@ -0,0 +1,31 @@
+    });
+  });
+
+  describe('a11y', () => {
```
I think these checks are already being done on line 14. There's no need to have this separate a11y test

katie-bobbe-genesys commented:
This is a minor comment, but if you don't mind could you change the titles of your PRs to a format like `feat(gux-card): addition of new card component`?

gavin-everett-genesys commented on src/components/beta/gux-card/gux-card.tsx:
```diff
@@ -0,0 +1,33 @@
+   * Card Type
+   */
+  @Prop()
+  type: GuxCardType = 'default';
```
Good spot! I will look into other property names that will suit it better.

gavin-everett-genesys commented on src/components/beta/gux-card/example.html:
```diff
@@ -0,0 +1,64 @@
+<gux-card-beta type="raised">
+  <div slot="content">raised card</div>
+</gux-card-beta>
+<h2>Flexible Card</h2>
```
Yes so the card itself can grow and shrink depending on the content it receives. As far a I am aware this applies to all cards so the heading "Flexible Card" will cause confusion. I will update this to something more meaningful.

gavin-everett-genesys commented:
@daragh-king-genesys @katie-bobbe-genesys @MattCheely 
Hi all, The following changes have been made to the card component in relation to some of the PR comments above.

- Renamed 'type' prop to 'emphasis' as I think this would represent the styling better. Also renamed the 'default' styling to 'outline' same as the design docs.
- Removed the named slot like Katie said since the slot will take in all the child content from container so a named slot is not needed.
- Updated the examples page to display the different variations (Potentially add use case section?)
- Removed unnecessary content div.
- Updated tests to reflect the changes above.

Changes can be viewed here shortly:
https://apps.inindca.com/common-ui-docs/genesys-webcomponents/feature/COMUI-886/#gux-card

Design Docs:
https://spark.genesys.com/7978beca0/p/493bf5-card
https://www.figma.com/file/vQxcSpzB12t4a47IOb6Gua/Spark---Component-Library?node-id=3923%3A24586

katie-bobbe-genesys commented on src/components/beta/gux-card/gux-card.tsx:
```diff
@@ -0,0 +1,28 @@
+   * Card Type
+   */
+  @Prop()
+  emphasis: GuxCardType = 'outline';
```
I'm not sure, but maybe `accent` would be a slightly better prop name? We use `accent` in our `gux-button` component for the different styles. `emphasis` gives me the impression that the different types convey importance of the element. Not sure though

gavin-everett-genesys replied:
I used the word emphasis from the design doc. Cards with the filled option result in less emphasis than raised or outlined cards and are used to indicate secondary or nested content whereas raised and outlined are for primary content. Accent also sounds good.

katie-bobbe-genesys commented on src/components/beta/gux-card/gux-card.tsx:
```diff
@@ -0,0 +1,28 @@
+})
+export class GuxCard {
+  /**
+   * Card Type
```
Instead of `Card Type` you change this description to be a bit more specific?

katie-bobbe-genesys reviewed (commented):
I had a few minor comments, but once they are resolved I will approve this PR 👍

katie-bobbe-genesys commented on src/components/beta/gux-card/gux-card.tsx:
```diff
@@ -0,0 +1,33 @@
+
+  render(): JSX.Element {
+    return (
+      <div class={`gux-card gux-${this.type}`}>
```
I looked into this and I don't think the card needs an aria role

daragh-king-genesys commented on src/components/beta/gux-card/gux-card.tsx:
```diff
@@ -0,0 +1,24 @@
+
+import { GuxCardEmphasis } from './gux-card.types';
+
+@Component({
```
You can add a Comment here so that the slot gets documented in the readme

have a look at `gux-tooltip` for an example

daragh-king-genesys commented on src/components/beta/gux-card/gux-card.tsx:
```diff
@@ -0,0 +1,24 @@
+      <div class={`gux-card gux-${this.emphasis}`}>
+        <slot />
+      </div>
+    ) as JSX.Element;
```
Generally I try to avoid unnecessary containers. You can just return `<slot />` here and use attribut selectors on the host in your css file to get the same result. Something like
```
:host {
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  width: fit-content;
  padding: @spacing-medium;
  background-color: @gux-grey-100;
  border: 1px solid @gux-grey-60;
  border-radius: @spacing-2xs;
  background-color: @gux-grey-100;
}

:host([emphasis="default"]) {
  background-color: @gux-grey-100;
}

:host([emphasis="filled"]) {
  background-color: @gux-grey-90;
}

:host([emphasis="raised"]) {
  .gux-shadow-10();
}
```

daragh-king-genesys commented on src/components/beta/gux-card/tests/gux-card.e2e.ts:
```diff
@@ -0,0 +1,21 @@
+import { newSparkE2EPage, a11yCheck } from '../../../../../tests/e2eTestUtils';
+
+describe('gux-card-beta', () => {
+  describe('#render different card variations eg(outline,outline(explicit),filled,raised)', () => {
```
I don't see the need for this describe block

gavin-everett-genesys replied:
Yep will do, I had an a11y check underneath but moved it into the the first describe.

daragh-king-genesys commented on src/components/beta/gux-card/tests/gux-card.spec.ts:
```diff
@@ -0,0 +1,91 @@
+const language = 'en';
+
+describe('gux-card-beta', () => {
+  describe('#render outline card', () => {
```
For these standard prop snapshot tests I would just use the array `it(`should render component as expected (${index + 1})`, async () => {` approach like you used in the e2e tests

gavin-everett-genesys commented:
@daragh-king-genesys @katie-bobbe-genesys @MattCheely 

Addressed PR comments above. Updates include

- Update styling property of card from `emphasis` to `accent`
- Addition of usage tracking.
- Added comment for slot.
- Utilised Host instead of unnecessary container.
- Removed wrapper describe block in e2e and spec.

View changes below,
https://apps.inindca.com/common-ui-docs/genesys-webcomponents/feature/COMUI-886/#gux-card
